### ISSUE TITLE
Container restart requires new rules

### ISSUE THREAD

**Issue by figassis** [chaifeng/ufw-docker#31]
Restarting a container (system reboot, docker restart, etc) causes containers to start with different ips and so the rules stop working. This means that if a service fails and has to restart, unsupervised, this will cause downtime.

**Comment by ekaiser-uptycs** [comment]
the rules cover 192.168.0.0 to 192.168.255.255, 10.0.0.0 to 10.255.255.255, and 172.16.0.0 to 172.31.255.255 - when you say the rules stop working, do you mean the UFW rules you're putting in place to allow access to container ports from outside the host?

**Comment by shinebayar-g** [comment]
See: https://github.com/chaifeng/ufw-docker/issues/22#issuecomment-700452102

**Comment by shinebayar-g** [comment]
Related: https://github.com/chaifeng/ufw-docker/issues/36

**Comment by figassis** [comment]
I've decided to simply bind containers to host IPs:
```
ports:
	- 127.0.0.1:80:80
```

**Comment by shinebayar-g** [comment]
If your server doesn't need to be public you don't need to expose it to public port at first place.

**Comment by figassis** [comment]
Yes, but even when it's not public and you expose it to say :3000, it will be accessible publicly disregarding UFW, which many many many developers are unaware. In fact, the bind to IP is also not very well known, so you see people stuck trying to find a solution.

**Comment by shinebayar-g** [comment]
You have to know `-p 80:80` means bind to EVERY SINGLE IP on the host. You have to be specific about that, like you mentioned bind only on `127.0.0.1` If many developers aren't aware of `127.0.0.1` god bless them.

**Comment by figassis** [comment]
You don't have to have auth on every internal service, but not exposing every internal service tot he internet is one of the first requirements.

I believe this is due to the fact that most tutorials I've seen never touching on the local vs public binding issue. They simply go, here's how you deploy an app via docker, and if you need to protect it, use the amazing UFW! I believe most of these authors are also unaware, I was for a while until I dove into the docs for this specific issue.

Moreover, people tend to assume behavior is standard across technologies. If you bind a web server to a public ip and then use UFW, it works. Docker is I believe the only outlier, so I can't blame people for expecting the same behavior from docker.

Finally, some people are trying to keep development environments as close to production, so that the same docker-compose file works in both, and assume that public binding will work in prod, and to protect dev we just need to add UFW to the host. This is a reasonable assumption, but is wrong. Devs are to blame, docker is to blame. I don't think UFW is to blame as docker is the one quietly overriding iptables and washing their hands.

**Comment by shinebayar-g** [comment]
Okay, I hacked together some crap and got working code, here is the [repo](https://github.com/shinebayar-g/ufw-docker-automated). All you have to do is run the container with special label `UFW_MANAGED=TRUE`. Please let me know your feedbacks. Of course code is crap, but hopefully it works.

**Comment by real-or-random** [comment]
If there's only one container that listens on the desired port, a simple workaround is a manual rule as described in the README:


> If you want to allow public networks to access the services provided by the Docker container, for example, the service port of a container is `80`. Run the following command to allow the public networks to access this service:
>
>     ufw route allow proto tcp from any to any port 80
>
> This allows the public network to access all published ports whose container port is `80`.
>
> Note: If we publish a port by using option `-p 8080:80`, we should use the container port `80`, not the host port `8080`.

It's just that the `ufw-docker` script doesn't accept `any` instead of a container name. @chaifeng Do you think that supporting `any` will be a good addition?

**Comment by danutzzzzz** [comment]
ok that's useful but crucially doesn't account if a container changes IP during operational status.

**Comment by thiswillbeyourgithub** [comment]
I resorted to adding a crontab to run `ufw-docker reload` every 10 minutes. This way if I forget to reload them and add a new container it will be outdated at most 10 minutes.

Right?

**Comment by danutzzzzz** [comment]
not sure what you mean.. i resorted to statically assigning the containers to an IP that i needed to ensure always remained accessible. 

Im hoping Dev will come up with a solution to monitor container IP's and have the rules follow.

**Comment by thiswillbeyourgithub** [comment]
I'm actually surprised no solution came up besides the daemon from @shinebayar-g .

It really seems like we could *just*:
- Store container names and rule content in a toml.
- Recreate the rules after boot.
- Use a daemon that checks wether the mapping "container name" to "ip adresses" has changed. If so we reload the rules.

Couldn't this just be an easy-to-maintain, working-everywhere python script? What complexity am I missing here? The aforementioned daemon has to be compiled (my go version is not appropriate) and needs maintenance apparently.

**Comment by shinebayar-g** [comment]
> Im hoping Dev will come up with a solution to monitor container IP's and have the rules follow.

It's not an official solution. But this project https://github.com/shinebayar-g/ufw-docker-automated does exactly that.
I've some features planned like
- Docker v29 compatibility
- File based configuration on top of the label based configuration. Because editing the running container's label requires container restart which is not ideal for some workloads.
- Adding CLI helper to see if all the containers are properly secured etc..

**Comment by shinebayar-g** [comment]
@thiswillbeyourgithub imho this project's goal is to actually provide the secure firewall rules to fix the "whole". In that regard it does the job perfectly well and I'm happy with the results. It even provides installer script which is awesome.

However, running a daemon service to monitor docker containers and applying rules etc could be definitely out of scope of the project.

> Couldn't this just be an easy-to-maintain, working-everywhere python script?

My initial version was exactly that, python script. But in order to connect to the Docker Engine, you need a client SDK to listen to the events and do various operations. However, to run Python, you need to have Python installed, then docker sdk dependencies installed etc.. As a good python user, you also need to have a virtualenv set up, otherwise docker dependencies can conflict with the system dependencies.
For this reason, I ditched Python and rewritten in Golang. With Golang, you just need **one file**. Copy that anywhere in your server and you're golden.

**Comment by thiswillbeyourgithub** [comment]
Seems like fair points in 2020 when you stated the daemon but nowadays with uv i'm not sure it holds up anymore tbh.

And actually I don't think you really need a docker client api. With subprocess you can just use `docker ps --format json` to get the container's name and associated network, then `docker network inspect [name] --format json` to get the ip of a container. This could even run every minute because you just need the `ps` output to know if the container has been restarted recently.

Overall this looks easier than installing the appropriate golang version without breaking anything.

**Comment by shinebayar-g** [comment]
You don't need Golang to run the binary?

**Comment by thiswillbeyourgithub** [comment]
Absolutely but then the user depends on you to do a release. I really don't want to sound entitled, it's of course at your own pace. But I think it's a good thing to not depend on anyone's pace :)
